### PR TITLE
Default Paragraph Behavior

### PR DESCRIPTION
**Pull request by tadean**
This PR adds a default paragraph behavior plugin to serve as the base for future modification.

## Description
This pull request adds a default paragraph behavior plugin (**AZDefaultParagraphBehavior**) to serve as a base class for paragraph styles. This is intended to contain the global paragraph styles that are potentially available to all paragraphs.
Currently it handles adding a default bottom spacing element to paragraphs.

Previous functionality that handled loading paragraph-specific css and js libraries has been refactored and placed into
this plugin as well.

The new behavior is enabled on the existing **az_text** paragraph.

The existing paragraph test is altered to check that the plugin is working.

Contributing documentation is updated.

## Related Issue
#331 

## How Has This Been Tested?
Create a text paragraph. Verify that the Additional Options detail pane exists and contains the Bottom Spacing dropdown.Verify that the selected spacing element is added to the paragraph in rendering.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I've added this Pull Request to the AZ Quickstart project in the right column.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by tadean** [comment]
image

image

**Review by joeparsons — APPROVED** [review]

**Review by mmunro-ltrr — APPROVED** [review]
After messing up the live demo, it's only polite to approve this!

**Review by jwmoore1 — APPROVED** [review]

File: modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZDefaultParagraphBehavior.php [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+   * Drupal\Core\Asset\LibraryDiscoveryInterface definition.
+   *
+   * @var \Drupal\Core\Asset\LibraryDiscoveryInterface
+   */
+  protected $libraryDiscovery;
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZDefaultParagraphBehavior.php:48]: I think the preferred way of injecting additional dependencies (beyond what the parent class does is to actually override the `__construct()` method with additional typeHinted dependencies in the method signature, and then inject all of the dependencies (both those defined by the parent and the additional ones defined in the extended class) in the `create()` method.  We have an example of this in the az_quickstart_core module here:
https://github.com/az-digital/az_quickstart/blob/main/modules/custom/az_core/src/Form/QuickstartCoreSettingsForm.php#L31-L57

**Review by joeparsons — CHANGES_REQUESTED** [review]

**Comment by camikazegreen** [comment]
In the help text for the Bottom Spacing, can we link directly to: https://digital.arizona.edu/arizona-bootstrap/docs/2.0/utilities/spacing

**Comment by joeparsons** [comment]
This looks AMAZING overall!

File: modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZDefaultParagraphBehavior.php [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+ * Provides a base behavior for Quickstart Paragraph Defaults.
+ *
+ * @ParagraphsBehavior(
+ *   id = "az_default_paragraph_behavior",
+ *   label = @Translation("Default Quickstart Paragraph Behavior"),
+ *   description = @Translation("Provides default Quickstart paragraph settings."),
+ *   weight = 0
+ * )
+ */
+class AZDefaultParagraphBehavior extends ParagraphsBehaviorBase {
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZDefaultParagraphBehavior.php:22]: Nitpick:  Would it be better to name our behavior `AZDefaultParagraphsBehavior` (with an `s`).  It seems that within the Paragraphs module, behaviors are always referred to as Paragraphs Behaviors.

**Review by trackleft — APPROVED** [review]
Magic!

File: modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZDefaultParagraphBehavior.php (hunk 1 above)
Comment (joeparsons) [reply modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZDefaultParagraphBehavior.php:48]: @tadean This documentation on drupal.org provides some similar examples: https://www.drupal.org/docs/8/api/services-and-dependency-injection/dependency-injection-for-a-form

Comment (tadean) [reply modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZDefaultParagraphBehavior.php:78]: Thanks @joeparsons ! Updated to the style of overriding that overrides the constructor.

File: modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZDefaultParagraphBehavior.php (hunk 2 above)
Comment (tadean) [reply modules/custom/az_paragraphs/src/Plugin/paragraphs/Behavior/AZDefaultParagraphBehavior.php]: I agree, that seems better. Updated to change the name to `AZDefaultParagraphsBehavior`.

**Comment by tadean** [comment]
Updated with the following:

- Remove `az_` scoping on FAPI elements inside the `az_display_settings` element.
- Update style of container dependency injection.
- Remove a bare call to `t()`
- Added link to spacing documentation
- Changed name of class to `AZDefaultParagraphsBehavior`

**Review by joeparsons — APPROVED** [review]

**Review by trackleft — APPROVED** [review]

**Pull request merged**